### DISCUSSION TITLE
Running in @QuarkusTest fails with ContextNotActiveException

### DISCUSSION THREAD

**Discussion by tomsontom** [quarkusio/quarkus#38727] [Q&A]
I've a base class like this:
```java
public abstract class BaseReadonlyHandler extends BaseHandler {

    public BaseReadonlyHandler(EntityManager em) {
        super(em);
        // Fails with @QuarkusTest
        this.em.unwrap(Session.class).setDefaultReadOnly(true);
    }
}

@Singleton
public class MyHandler {
  @Inject
  public MyHandler(EntityManager em) {
    super(em);
  }

  // ...
}
```
Starting quarkus in Dev-Mode all is fine but when trying to write a JUnit-Tests like this:
```java
@QuarkusTest
public class MyHandlerTest {
  @Inject
  public MyHandler handler;
}
```
leading to the following exception
```
Caused by: jakarta.enterprise.context.ContextNotActiveException: Cannot use the EntityManager/Session because neither a transaction nor a CDI request context is active. Consider adding @Transactional to your method to automatically activate a transaction, or @ActivateRequestContext if you have valid reasons not to use transactions.
        at io.quarkus.hibernate.orm.runtime.session.TransactionScopedSession.acquireSession(TransactionScopedSession.java:107)
        at io.quarkus.hibernate.orm.runtime.session.TransactionScopedSession.setDefaultReadOnly(TransactionScopedSession.java:671)
        at org.hibernate.engine.spi.SessionLazyDelegator.setDefaultReadOnly(SessionLazyDelegator.java:157)
```
Is there anything I can do? My current workaround is to put the EM into-Readonly-Model upon first access, which is not too bad but it feels wrong to just make the tests succeed.

**Comment by rikhmaier** [comment]
The problem is not specific to `@QuarkusTest`; it is caused by using the injected `EntityManager` from a bean constructor.

The injected `EntityManager`/`Session` is a proxy. Calling:

```java
em.unwrap(Session.class).setDefaultReadOnly(true);
```

requires an active transaction or request context, but neither is guaranteed while CDI is constructing `MyHandler`. Adding `@Transactional` to the test method will not fix that constructor call, because the bean must be created before the test method can run.

Move the Hibernate interaction to the transactional operation that uses the session:

```java
@Singleton
public class MyHandler extends BaseReadonlyHandler {

    @Inject
    public MyHandler(EntityManager em) {
        super(em);
    }

    @Transactional
    public Result handle(...) {
        em.unwrap(Session.class).setDefaultReadOnly(true);
        // execute the queries here
    }
}
```

If the handler has several entry points, put that setup in a shared method, but call it from inside each transactional entry point. `@PostConstruct` has the same lifecycle problem as the constructor, so it is not a reliable replacement.

For a genuinely non-transactional read operation, a method annotated with `@ActivateRequestContext` can activate the request context:

```java
@ActivateRequestContext
public Result handle(...) {
    em.unwrap(Session.class).setDefaultReadOnly(true);
    // read-only work
}
```

However, a transaction is generally preferable for database reads as well, because all reads then use one consistent database transaction.

Your lazy “set it on first access” workaround succeeds because the session is accessed after a transaction/request context has already been established. I would keep the same timing but make it explicit at the business-method boundary instead of hiding it behind lazy initialization.

Quarkus documentation also states that the injected `EntityManager`/`Session` proxy requires an active transaction for interaction; read-only access without a transaction is only supported while a request context is active:
https://quarkus.io/guides/hibernate-orm/#cdi-integration
